src/webrtc/groupCall.ts

```diff
@@ -0,0 +1,762 @@
+                return;
+            }
+        }
+
+        // Only initiate a call with a user who has a userId that is lexicographically
+        // less than your own. Otherwise, that user will call you.
+        if (member.userId < localUserId) {
+            return;
+        }
+
+        const call = this.client.createCall(this.room.roomId, member.userId);
+
+        let callPromise;
+
+        if (this.type === CallType.Video) {
+            callPromise = call.placeVideoCall();
+        } else {
+            callPromise = call.placeVoiceCall();
+        }
+
+        callPromise.then(() => {
+            if (this.dataChannelsEnabled) {
+                call.createDataChannel("datachannel", this.dataChannelOptions);
+            }
+        });
+
+        if (participant) {
+            participant.replaceCall(call, sessionId);
+        } else {
+            participant = new GroupCallParticipant(
+                this,
+                member,
+                sessionId,
+                call,
+            );
+            // TODO: Should we wait until the call has been answered to push the participant?
+            // Or do we hide the participant until their stream is live?
+            // Does hiding a participant without a stream present a privacy problem because
+            // a participant without a stream can still listen in on other user's streams?
+            this.participants.push(participant);
+            this.reEmitter.reEmit(participant, Object.values(GroupCallParticipantEvent));
+            this.emit(GroupCallEvent.ParticipantsChanged), this.participants;
+        }
+    };
+
+    onActiveSpeakerLoop = () => {
+        let topAvg;
+        let nextActiveSpeaker;
+
+        for (const participant of this.participants) {
+            let total = 0;
+
+            for (let i = 0; i < participant.activeSpeakerSamples.length; i++) {
+                const volume = participant.activeSpeakerSamples[i];
+                total += Math.max(volume, SPEAKING_THRESHOLD);
+            }
+
+            const avg = total / ACTIVE_SPEAKER_SAMPLES;
+
+            if (!topAvg || avg > topAvg) {
+                topAvg = avg;
+                nextActiveSpeaker = participant.member;
+            }
+        }
+
+        if (nextActiveSpeaker && topAvg > SPEAKING_THRESHOLD) {
+            if (nextActiveSpeaker && this.activeSpeaker !== nextActiveSpeaker) {
+                this.activeSpeaker.activeSpeaker = false;
+                nextActiveSpeaker.activeSpeaker = true;
+                this.activeSpeaker = nextActiveSpeaker;
+                this.emit(GroupCallEvent.ActiveSpeakerChanged, this.activeSpeaker);
+            }
+        }
+
+        this.activeSpeakerLoopTimeout = setTimeout(
+            this.onActiveSpeakerLoop,
+            ACTIVE_SPEAKER_INTERVAL,
+        );
+    };
+
+    /**
+     * Utils
+     */
+
+    // TODO: move this elsewhere or get rid of the retry logic. Do we need it?
+    sendStateEventWithRetry(
```

Comment:
I don't actually know that we still want this for this PR. I don't think it was actually making calls any more stable, but I don't know for sure.